src/proxy/connection_manager.rs

```diff
@@ -82,19 +81,29 @@ pub struct ConnectionGuard {
     watch: Option<DrainWatcher>,
 }
 
-impl ConnectionGuard {
-    pub async fn handle_connection(
-        mut self,
-        send: impl Future<Output = Result<(), Error>> + Sized,
-    ) -> Result<(), Error> {
-        let watch = self.watch.take().expect("watch cannot be taken twice");
+// For reasons that I don't fully understand, this uses an obscene amount of stack space when written as a normal function,
+// amounting to ~1kb overhead per connection.
+// Inlining it removes this entirely, and the macro ensures we do it consistently across the various areas we use it.
+#[macro_export]
+macro_rules! handle_connection {
```

Comment:
What happens with `#[inline(always)]` rather than going all the way to a macro?

Reply:
:eyes: